lib/srv/app/aws/handler.go

```diff
@@ -17,211 +17,171 @@ limitations under the License.
+func (s *signerHandler) serveHTTP(w http.ResponseWriter, req *http.Request) error {
+	sessCtx, err := common.GetSessionContext(req)
 	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	// emit audit event with original request, but change the URL since we resolved and rewrote it.
-	signedReq.Body = io.NopCloser(bytes.NewReader(payload))
-	if isDynamoDBEndpoint(resolvedEndpoint) {
-		err = sessionCtx.Audit.OnDynamoDBRequest(req.Context(), sessionCtx, signedReq, resp, resolvedEndpoint)
-	} else {
-		err = sessionCtx.Audit.OnRequest(req.Context(), sessionCtx, signedReq, resp, resolvedEndpoint)
+		return trace.Wrap(err)
 	}
+
+	// It's important that we resolve the endpoint before modifying the request headers,
+	// as they may be needed to resolve the endpoint correctly.
+	re, err := resolveEndpoint(req)
 	if err != nil {
-		s.Log.WithError(err).Warn("Failed to emit audit event.")
+		return trace.Wrap(err)
 	}
-	return resp, nil
-}
 
-func (s *SigningService) formatForwardResponseError(rw http.ResponseWriter, r *http.Request, err error) {
-	common.SetTeleportAPIErrorHeader(rw, err)
-
-	switch trace.Unwrap(err).(type) {
-	case *trace.BadParameterError:
-		s.Log.Debugf("Failed to process request: %v.", err)
-		rw.WriteHeader(http.StatusBadRequest)
-	case *trace.AccessDeniedError:
-		s.Log.Infof("Failed to process request: %v.", err)
-		rw.WriteHeader(http.StatusForbidden)
-	default:
-		s.Log.Warnf("Failed to process request: %v.", err)
-		rw.WriteHeader(http.StatusInternalServerError)
+	// rewrite headers before signing the request to avoid signature validation problems.
+	unsignedReq, err := rewriteRequest(req, re)
+	if err != nil {
+		return trace.Wrap(err)
 	}
-}
 
-// prepareSignedRequest creates a new HTTP request and rewrites the header from the original request and returns a new
-// HTTP request signed by STS AWS API.
-func (s *SigningService) prepareSignedRequest(r *http.Request, payload []byte, re *endpoints.ResolvedEndpoint, sessionCtx *common.SessionContext) (*http.Request, error) {
-	url, err := urlForResolvedEndpoint(r, re)
+	signedReq, err := s.SignRequest(unsignedReq,
+		&awsutils.SigningCtx{
+			SigningName:   re.SigningName,
+			SigningRegion: re.SigningRegion,
+			Expiry:        sessCtx.Identity.Expires,
+			SessionName:   sessCtx.Identity.Username,
+			AWSRoleArn:    sessCtx.Identity.RouteToApp.AWSRoleARN,
+			AWSExternalID: sessCtx.App.GetAWSExternalID(),
+		})
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return trace.Wrap(err)
+	}
+	recorder := httplib.NewResponseStatusRecorder(w)
+	s.fwd.ServeHTTP(recorder, signedReq)
+
+	var auditErr error
+	if isDynamoDBEndpoint(re) {
+		auditErr = sessCtx.Audit.OnDynamoDBRequest(unsignedReq.Context(), sessCtx, unsignedReq, recorder.Status(), re)
+	} else {
+		auditErr = sessCtx.Audit.OnRequest(unsignedReq.Context(), sessCtx, unsignedReq, recorder.Status(), re)
 	}
-	reqCopy, err := http.NewRequest(r.Method, url, bytes.NewReader(payload))
 	if err != nil {
```

Comment:
good catch!

Reply:
I've also added `closeContext` to the handler since a cancelled request context should not affect audit log event emitting. It gets the close context from the app server, just like we do in the Azure handler.